astropy/io/fits/header.py

```diff
@@ -1044,132 +1026,48 @@ def update(self, *args, **kwargs):
-        # arguments, but realistically speaking that use case will not come up
-
-        # The fact that Python is "flexible" in allowing positional args to be
-        # passed in as keyword args makes this a little more complicated than
-        # it otherwise would be :/
-        issubset = set(kwargs).issubset(set(legacy_args))
-        if (len(args) >= 2 or
-            (len(args) == 1 and 'value' in kwargs and issubset) or
-            (len(args) == 0 and 'key' in kwargs and 'value' in kwargs and
-             issubset)):
-            # This must be a legacy update()
-            warnings.warn(
-                "The use of header.update() to add new keywords to a header "
-                "is deprecated.  Instead, use either header.set() or simply "
-                "`header[keyword] = value` or "
-                "`header[keyword] = (value, comment)`.  header.set() is only "
-                "necessary to use if you also want to use the before/after "
-                "keyword arguments.", AstropyDeprecationWarning)
-
-            for k, v in zip(legacy_args, args):
-                if k in kwargs:
-                    raise TypeError(
-                        '%s.update() got multiple values for keyword '
-                        'argument %r' % (self.__class__.__name__, k))
-                kwargs[k] = v
-
-            keyword = kwargs.get('key')
-            value = kwargs.get('value')
-            comment = kwargs.get('comment')
-            before = kwargs.get('before')
-            after = kwargs.get('after')
-            savecomment = kwargs.get('savecomment')
-
-            # Handle the savecomment argument which is not currently used by
-            # Header.set()
-            if keyword in self and savecomment:
-                comment = None
-
-            self.set(keyword, value, comment, before, after)
+        if args:
+            other = args[0]
         else:
-            # The rest of this should work similarly to dict.update()
-            if args:
-                other = args[0]
-            else:
-                other = None
+            other = None
 
-            def update_from_dict(k, v):
-                if not isinstance(v, tuple):
-                    card = Card(k, v)
-                elif 0 < len(v) <= 2:
-                    card = Card(*((k,) + v))
+        def update_from_dict(k, v):
+            if not isinstance(v, tuple):
+                card = Card(k, v)
+            elif 0 < len(v) <= 2:
+                card = Card(*((k,) + v))
+            else:
+                raise ValueError(
+                    'Header update value for key %r is invalid; the '
+                    'value must be either a scalar, a 1-tuple '
+                    'containing the scalar value, or a 2-tuple '
+                    'containing the value and a comment string.' % k)
+            self._update(card)
+
+        if other is None:
+            pass
+        elif hasattr(other, 'iteritems'):
```

Comment:
I agree it could be changed but I prefer not to do this in this PR.
And anyway on python 3 it works with the next `elif hasattr(other, 'keys')`.